src/lib/input/input.ts

```diff
@@ -170,7 +170,20 @@ export class MatInput extends _MatInputMixinBase implements MatFormFieldControl<
    * @docs-private
    */
   @Input()
-  get required(): boolean { return this._required; }
+  get required(): boolean {
+    if (this._required) {
+      return this._required;
+    }
+
+    // The required attribute is set
+    // when the control return an error from validation with an empty value
+    if (this.ngControl && this.ngControl.control && this.ngControl.control.validator) {
+      const emptyValueControl = Object.assign({}, this.ngControl.control);
+      (emptyValueControl as any).value = null;
+      return 'required' in (this.ngControl.control.validator(emptyValueControl) || {});
```

Comment:
This is somewhat wasteful, because it'll run on every change detection run. Wouldn't it be enough to only check it on init?

Reply:
I thought about it already, the thing is what will happen if we (re)set the formControl validators after the init ? I never found a case where I needed to do this but can we ignored it ?

Reply:
> Could the input just check for the presence of the `RequiredValidator` directive?

I possibly didn't understand but, checking the required directive was already done. The change is about checking if the ValidatorFn of the formControl make the field required without checking it. It's only for reactive forms, without using the directive on the template.

I thought about another way to do it, it's to only check if control.errors contains 'required' field.
But I don't know if it's a better way.